### ISSUE TITLE
serialise error

### ISSUE THREAD

**Issue by lijing28101** [EI-CoreBioinformatics/mikado#269]
Hi, I'm running mikado serialise on version 2.
Here is the error
```
mikado serialise --start-method spawn --single-thread --blast_targets ${genome} --json-conf configuration.yaml --xml ${blastxml} --orfs ${orfs}
2020-02-14 10:51:09,318 - main - __init__.py:120 - ERROR - main - MainProcess - Mikado crashed, cause:
2020-02-14 10:51:09,318 - main - __init__.py:121 - ERROR - main - MainProcess - ((278, 278), (440101, 439535), '|||||||||||||||  ||||| | |||||||  ||| |||||||||||||||||||||||||||||| |||||  ||| |||||||||| ||||||||||||||||||||||||||| ||||||| |||||| |||||||||   |  |||  |       |||||||||||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||||||   |||||||| |||||||| |||||||||||||||||||||| |||||||', 'ACACTACCCTAACACAGCCCTAATCTAACCCTGGCCAACCTGTCTCTCAACTTACCCTCCATTACCCTGCCTCC--ACTCGTTACCCTGTCCCATTCAACCATACCACTCCGAACCACCATCCATCCCTCTACTTACTACCACT--C--ACC--C-------ACCGTTACCCTCCAATTACCCATATCCAACCC-ACTGCCACTTACCCTACCATTACCCTACCATCCACCATGACCTACTCACCATACTGTTCTTCTACCCACCATATTGAAACGCTAACAAA', 'ACACTACCCTAACACTACCCTACTTTAACCCTAACCATCCTGTCTCTCAACTTACCCTCCATTACCCTTCCTCCCCACTTGTTACCCTGTTCCATTCAACCATACCACTCCGAACCACAATCCATCTCTCTACGTACTACCACCAACCCACCGTCCACCATAACCGTTACCCTCCAATTACCCATATCCAACCCCACTGCCACTTACCCTACCATTACCCTACCATCCACCATCTGCTACTCACTATACTGTTGTTCTACCCACCATATTGAAACGTTAACAAA')
Traceback (most recent call last):
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/__init__.py", line 106, in main
    args.func(args)
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/subprograms/serialise.py", line 380, in serialise
    load_blast(args, logger)
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/subprograms/serialise.py", line 124, in load_blast
    part_launcher(filenames)
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/subprograms/serialise.py", line 54, in xml_launcher
    xml_serializer()
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/serializers/blast_serializer/xml_serialiser.py", line 536, in __call__
    self.serialize()
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/serializers/blast_serializer/xml_serialiser.py", line 479, in serialize
    cache=cache, max_target_seqs=self.__max_target_seqs, logger=self.logger)
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/serializers/blast_serializer/xml_serialiser.py", line 703, in objectify_record
    **hit_dict_params)
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/serializers/blast_serializer/utils.py", line 161, in prepare_hit
    hsp_dict, ident, posit = prepare_hsp(hsp, counter, qmultiplier=qmulti, tmultiplier=tmulti)
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/serializers/blast_serializer/utils.py", line 37, in prepare_hsp
    match, identical_positions, positives = _prepare_aln_strings(hsp, qmultiplier=qmultiplier, tmultiplier=tmultiplier)
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/serializers/blast_serializer/utils.py", line 110, in _prepare_aln_strings
    hsp.match, hsp.query, hsp.sbjct)
AssertionError: ((278, 278), (440101, 439535), '|||||||||||||||  ||||| | |||||||  ||| |||||||||||||||||||||||||||||| |||||  ||| |||||||||| ||||||||||||||||||||||||||| ||||||| |||||| |||||||||   |  |||  |       |||||||||||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||||||   |||||||| |||||||| |||||||||||||||||||||| |||||||', 'ACACTACCCTAACACAGCCCTAATCTAACCCTGGCCAACCTGTCTCTCAACTTACCCTCCATTACCCTGCCTCC--ACTCGTTACCCTGTCCCATTCAACCATACCACTCCGAACCACCATCCATCCCTCTACTTACTACCACT--C--ACC--C-------ACCGTTACCCTCCAATTACCCATATCCAACCC-ACTGCCACTTACCCTACCATTACCCTACCATCCACCATGACCTACTCACCATACTGTTCTTCTACCCACCATATTGAAACGCTAACAAA', 'ACACTACCCTAACACTACCCTACTTTAACCCTAACCATCCTGTCTCTCAACTTACCCTCCATTACCCTTCCTCCCCACTTGTTACCCTGTTCCATTCAACCATACCACTCCGAACCACAATCCATCTCTCTACGTACTACCACCAACCCACCGTCCACCATAACCGTTACCCTCCAATTACCCATATCCAACCCCACTGCCACTTACCCTACCATTACCCTACCATCCACCATCTGCTACTCACTATACTGTTGTTCTACCCACCATATTGAAACGTTAACAAA')
```
Could you help me to figure out the error?
Thanks,
Jing

**Comment by lucventurini** [comment]
Dear @lijing28101 , thank you for reporting. I am having a look now.

**Comment by lijing28101** [comment]
Thanks! @lucventurini 
I also tried to using the configure.yaml produced by version 1.2.4 and then run the mikado serialise 1.2.4. There is another error:
```
mikado serialise --start-method spawn --single-thread --blast_targets ${genome} --json-conf configuration.yaml --xml ${blastxml} --orfs ${orfs}
/work/LAS/mash-lab/jing/bin/Anaconda3/envs/pyrpipe_human/lib/python3.6/site-packages/Mikado/configuration/configurator.py:529: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  scoring = yaml.load(scoring_file)
2020-02-14 10:38:27,673 - main - __init__.py:124 - ERROR - main - MainProcess - Mikado crashed, cause:
2020-02-14 10:38:27,673 - main - __init__.py:125 - ERROR - main - MainProcess - No transaction is begun.
Traceback (most recent call last):
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/pyrpipe_human/lib/python3.6/site-packages/Mikado/__init__.py", line 110, in main
    args.func(args)
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/pyrpipe_human/lib/python3.6/site-packages/Mikado/subprograms/serialise.py", line 323, in serialise
    load_orfs(args, logger)
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/pyrpipe_human/lib/python3.6/site-packages/Mikado/subprograms/serialise.py", line 144, in load_orfs
    serializer()
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/pyrpipe_human/lib/python3.6/site-packages/Mikado/serializers/orf.py", line 314, in __call__
    self.serialize()
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/pyrpipe_human/lib/python3.6/site-packages/Mikado/serializers/orf.py", line 288, in serialize
    self.session.commit()
  File "/work/LAS/mash-lab/jing/bin/Anaconda3/envs/pyrpipe_human/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1034, in commit
    raise sa_exc.InvalidRequestError("No transaction is begun.")
sqlalchemy.exc.InvalidRequestError: No transaction is begun.
```

**Comment by lucventurini** [comment]
Dear @lijing28101 

I have updated the code that was failing in a separate branch, `issue-280`, just yesterday.
We should be merging it back to `master` within the week; I hope that it will solve your problems.

Many thanks for your patience.

**Comment by lucventurini** [comment]
Dear @lijing28101 

Another piece of information - which version and type of BLAST did you use? I do not see the pipe characters ("|") any longer in the output of NCBI BLASTX (I just tried).

Regardless, I completely changed that function yesterday, so hopefully this time Mikado will not crash.

I am linking this with #280 .

**Comment by lijing28101** [comment]
Hi @lucventurini ,

I used blast 2.7. 
When I tried the same data in different hpc, I didn't get the same error.
I guess the error maybe caused by the computer system? But I install all the dependency from your environment.yaml. I don't know why the error only happened in one hpc cluster.

Thanks,
Jing

**Comment by lucventurini** [comment]
Hi @lijing28101 
It is quite puzzling! I really do not have an answer to this, but I'm glad to hear that in the end you managed to get it running.

If that's OK with you, may I ask to repeat the run after we merge `issue-280` into `master `, probably tomorrow?

**Comment by lijing28101** [comment]
Hi @lucventurini 
Please tell me after you finish merging.
I can try the new version with another data.

**Comment by lucventurini** [comment]
This should now be fixed in version 2.0rc2 on both [PyPI](https://pypi.org/project/Mikado/2.0rc2/) and [BioConda](https://anaconda.org/bioconda/mikado). Please repoen if the issue represents itself with the new version!
